File: internal/catalog/catalog.go
package catalog

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var (
	notebookTablePattern   = regexp.MustCompile(`(?m)^\| \[(?P<name>[^\]]+)\]\([^)]*\) \| (?P<description>.+?) \|$`)
	topLevelHeadingPattern = regexp.MustCompile(`(?m)^##\s+(?P<title>.+)$`)
)

var scriptSummaries = map[string]string{
	"check_embeddings_matrix.py": "Generate BERT embeddings for a custom sentence and inspect raw tensor values.",
	"check_attention_heads.py":   "Visualize multi-head attention heatmaps for a sample sentence using BERT attention outputs.",
	"manim_aggregation.py":       "Render the predictive ML data aggregation animation with Manim.",
	"manim_bpe_tokenization.py":  "Render the BPE tokenization animation used in the workshop.",
	"manim_predictive_ml.py":     "Render the predictive machine learning explainer animation with Manim.",
	"manim_quantization.py":      "Render the quantization explainer animation with Manim.",
}

type CommandSpec struct {
	Label   string `json:"label"`
	Command string `json:"command"`
}

type Resource struct {
	Kind       string            `json:"kind"`
	Slug       string            `json:"slug"`
	Title      string            `json:"title"`
	Path       string            `json:"path"`
	Summary    string            `json:"summary"`
	Tags       []string          `json:"tags,omitempty"`
	Highlights []string          `json:"highlights,omitempty"`
	Commands   []CommandSpec     `json:"commands,omitempty"`
	Metadata   map[string]string `json:"metadata,omitempty"`
}

func (r Resource) SearchableText() string {
	parts := []string{r.Kind, r.Slug, r.Title, r.Path, r.Summary, strings.Join(r.Tags, " "), strings.Join(r.Highlights, " ")}
	return strings.ToLower(strings.Join(parts, "\n"))
}

type Stats struct {
	Sections  int `json:"sections"`
	Notebooks int `json:"notebooks"`
	Scripts   int `json:"scripts"`
	Videos    int `json:"videos"`
	Resources int `json:"resources"`
}

type Catalog struct {
	Root      string     `json:"root"`
	Sections  []Resource `json:"sections"`
	Notebooks []Resource `json:"notebooks"`
	Scripts   []Resource `json:"scripts"`
	Videos    []Resource `json:"videos"`
}

func (c Catalog) AllResources() []Resource {
	resources := make([]Resource, 0, len(c.Sections)+len(c.Notebooks)+len(c.Scripts)+len(c.Videos))
	resources = append(resources, c.Sections...)
	resources = append(resources, c.Notebooks...)
	resources = append(resources, c.Scripts...)
	resources = append(resources, c.Videos...)
	return resources
}

func (c Catalog) Search(query string, kinds ...string) []Resource {
	query = strings.TrimSpace(strings.ToLower(query))
	kindSet := map[string]struct{}{}
	for _, kind := range kinds {
		kindSet[kind] = struct{}{}
	}
	resources := c.AllResources()
	matches := make([]Resource, 0, len(resources))
	for _, resource := range resources {
		if len(kindSet) > 0 {
			if _, ok := kindSet[resource.Kind]; !ok {
				continue
			}
		}
		if query == "" || strings.Contains(resource.SearchableText(), query) {
			matches = append(matches, resource)
		}
	}
	return matches
}

func (c Catalog) Stats() Stats {
	return Stats{
		Sections:  len(c.Sections),
		Notebooks: len(c.Notebooks),
		Scripts:   len(c.Scripts),
		Videos:    len(c.Videos),
		Resources: len(c.Sections) + len(c.Notebooks) + len(c.Scripts) + len(c.Videos),
	}
}

func Build(root string) (Catalog, error) {
	readmePath := filepath.Join(root, "README.md")
	readmeBytes, err := os.ReadFile(readmePath)
	if err != nil {
		return Catalog{}, fmt.Errorf("read README: %w", err)
	}
	readme := string(readmeBytes)
	descriptions := parseNotebookDescriptions(readme)

	sections := parseReadmeSections(readme)
	notebooks, err := parseNotebooks(root, descriptions)
	if err != nil {
		return Catalog{}, err
	}
	scripts, err := parseScripts(root)
	if err != nil {
		return Catalog{}, err
	}
	videos, err := parseVideos(root)
	if err != nil {
		return Catalog{}, err
	}

	return Catalog{
		Root:      root,
		Sections:  sections,
		Notebooks: notebooks,
		Scripts:   scripts,
		Videos:    videos,
	}, nil
}

func parseNotebookDescriptions(readme string) map[string]string {
	matches := notebookTablePattern.FindAllStringSubmatch(readme, -1)
	result := map[string]string{}
	for _, match := range matches {
		if len(match) < 3 {
			continue
		}
		result[match[1]] = strings.TrimSpace(match[2])
	}
	return result
}

func parseReadmeSections(readme string) []Resource {
	matches := topLevelHeadingPattern.FindAllStringSubmatchIndex(readme, -1)
	resources := make([]Resource, 0, len(matches))
	for index, match := range matches {
		title := strings.TrimSpace(readme[match[2]:match[3]])
		if strings.EqualFold(title, "license") {
			continue
		}
		start := match[1]
		end := len(readme)
		if index+1 < len(matches) {
			end = matches[index+1][0]
		}
		body := strings.TrimSpace(readme[start:end])
		resources = append(resources, Resource{
			Kind:       "section",
			Slug:       slugify(title),
			Title:      title,
			Path:       "README.md",
			Summary:    extractFirstParagraph(body),
			Tags:       []string{"readme", "workshop", "concept"},
			Highlights: extractSectionHighlights(body),
		})
	}
	return resources
}

func parseNotebooks(root string, descriptions map[string]string) ([]Resource, error) {
	entries, err := os.ReadDir(filepath.Join(root, "notebooks"))
	if err != nil {
		return nil, fmt.Errorf("read notebooks: %w", err)
	}
	resources := make([]Resource, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() || filepath.Ext(entry.Name()) != ".ipynb" {
			continue
		}
		rel := filepath.ToSlash(filepath.Join("notebooks", entry.Name()))
		full := filepath.Join(root, rel)
		data, readErr := os.ReadFile(full)
		if readErr != nil {
			return nil, fmt.Errorf("read notebook %s: %w", rel, readErr)
		}
		headings := extractNotebookHeadings(data)
		title := strings.ReplaceAll(strings.TrimSuffix(entry.Name(), ".ipynb"), "_", " ")
		title = strings.Title(title)
		if len(headings) > 0 {
			title = strings.TrimSpace(strings.TrimLeft(headings[0], "# "))
		}
		summary := descriptions[entry.Name()]
		if summary == "" {
			summary = "Hands-on notebook for exploring the workshop topic."
		}
		highlights := []string{}
		if len(headings) > 1 {
			highlights = append(highlights, headings[1:]...)
		}
		resources = append(resources, Resource{
			Kind:       "notebook",
			Slug:       slugify(strings.TrimSuffix(entry.Name(), ".ipynb")),
			Title:      title,
			Path:       rel,
			Summary:    summary,
			Tags:       []string{"lab", "notebook", "jupyter"},
			Highlights: truncateSlice(highlights, 8),
			Commands: []CommandSpec{{
				Label:   "Launch notebook",
				Command: fmt.Sprintf("python3 -m jupyter notebook %s", rel),
			}},
		})
	}
	sortResources(resources)
	return resources, nil
}

func parseScripts(root string) ([]Resource, error) {
	entries, err := os.ReadDir(filepath.Join(root, "scripts"))
	if err != nil {
		return nil, fmt.Errorf("read scripts: %w", err)
	}
	resources := make([]Resource, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() || filepath.Ext(entry.Name()) != ".py" || entry.Name() == "names.py" {
			continue
		}
		stem := strings.TrimSuffix(entry.Name(), ".py")
		rel := filepath.ToSlash(filepath.Join("scripts", entry.Name()))
		summary := scriptSummaries[entry.Name()]
		if summary == "" {
			summary = humanizeScriptName(stem)
		}
		tags := []string{"script"}
		if strings.HasPrefix(entry.Name(), "manim_") {
			tags = append(tags, "animation")
		}
		if strings.HasPrefix(entry.Name(), "check_") {
			tags = append(tags, "inspection")
		}
		resources = append(resources, Resource{
			Kind:    "script",
			Slug:    slugify(stem),
			Title:   strings.Title(strings.ReplaceAll(stem, "_", " ")),
			Path:    rel,
			Summary: summary,
			Tags:    tags,
			Commands: []CommandSpec{{
				Label:   "Run script",
				Command: fmt.Sprintf("python3 %s", rel),
			}},
		})
	}
	sortResources(resources)
	return resources, nil
}

func parseVideos(root string) ([]Resource, error) {
	entries, err := os.ReadDir(filepath.Join(root, "video"))
	if err != nil {
		return nil, fmt.Errorf("read videos: %w", err)
	}
	resources := make([]Resource, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() || filepath.Ext(entry.Name()) != ".mp4" {
			continue
		}
		stem := strings.TrimSuffix(entry.Name(), ".mp4")
		resources = append(resources, Resource{
			Kind:    "video",
			Slug:    slugify(stem),
			Title:   strings.Title(strings.ReplaceAll(stem, "_", " ")),
			Path:    filepath.ToSlash(filepath.Join("video", entry.Name())),
			Summary: "Rendered workshop animation ready to watch locally or embed in talks.",
			Tags:    []string{"video", "animation"},
		})
	}
	sortResources(resources)
	return resources, nil
}

func extractNotebookHeadings(data []byte) []string {
	type notebookCell struct {
		CellType string            `json:"cell_type"`
		Source   []json.RawMessage `json:"source"`
	}
	type notebook struct {
		Cells []notebookCell `json:"cells"`
	}
	var parsed notebook
	if err := json.Unmarshal(data, &parsed); err != nil {
		return nil
	}
	headings := []string{}
	for _, cell := range parsed.Cells {
		if cell.CellType != "markdown" {
			continue
		}
		builder := strings.Builder{}
		for _, part := range cell.Source {
			var line string
			if err := json.Unmarshal(part, &line); err == nil {
				builder.WriteString(line)
			}
		}
		for _, line := range strings.Split(builder.String(), "\n") {
			trimmed := strings.TrimSpace(line)
			if strings.HasPrefix(trimmed, "#") {
				headings = append(headings, trimmed)
				break
			}
		}
	}
	return headings
}

func extractFirstParagraph(body string) string {
	lines := []string{}
	inCode := false
	for _, raw := range strings.Split(body, "\n") {
		line := strings.TrimSpace(raw)
		if strings.HasPrefix(line, "```") {
			inCode = !inCode
			continue
		}
		if inCode {
			continue
		}
		if line == "" {
			if len(lines) > 0 {
				break
			}
			continue
		}
		if strings.HasPrefix(line, "![") || strings.HasPrefix(line, "[![") || strings.HasPrefix(line, "|") || strings.HasPrefix(line, ">") || strings.HasPrefix(line, "### ") {
			if len(lines) > 0 {
				break
			}
			continue
		}
		lines = append(lines, line)
	}
	if len(lines) == 0 {
		return "Explore this part of the workshop from the terminal lab."
	}
	return strings.Join(lines, " ")
}

func extractSectionHighlights(body string) []string {
	lines := []string{}
	for _, raw := range strings.Split(body, "\n") {
		line := strings.TrimSpace(raw)
		if strings.HasPrefix(line, "### ") {
			lines = append(lines, strings.TrimSpace(strings.TrimPrefix(line, "### ")))
		}
	}
	return lines
}

func truncateSlice(values []string, max int) []string {
	if len(values) <= max {
		return values
	}
	return values[:max]
}

func humanizeScriptName(stem string) string {
	words := strings.ReplaceAll(stem, "_", " ")
	return fmt.Sprintf("Run the %s helper from the terminal lab.", words)
}

func slugify(value string) string {
	value = strings.ToLower(strings.TrimSpace(value))
	replacer := regexp.MustCompile(`[^a-z0-9]+`)
	value = replacer.ReplaceAllString(value, "-")
	return strings.Trim(value, "-")
}

func sortResources(resources []Resource) {
	sort.Slice(resources, func(i, j int) bool {
		return resources[i].Title < resources[j].Title
	})
}

File: src/data_in_ai_revolution/catalog.py
from __future__ import annotations

from dataclasses import dataclass, field
import json
import re
from pathlib import Path
from typing import Sequence


NOTEBOOK_TABLE_PATTERN = re.compile(
    r"^\| \[(?P<name>[^\]]+)\]\([^)]*\) \| (?P<description>.+?) \|$",
    re.MULTILINE,
)
TOP_LEVEL_HEADING_PATTERN = re.compile(r"^##\s+(?P<title>.+)$", re.MULTILINE)
SCRIPT_SUMMARIES: dict[str, str] = {
    "check_embeddings_matrix.py": "Generate BERT embeddings for a custom sentence and inspect the raw tensor values.",
    "check_attention_heads.py": "Visualize multi-head attention heatmaps for your own sample sentence using BERT attention outputs.",
    "manim_aggregation.py": "Render the predictive ML data aggregation animation with Manim.",
    "manim_bpe_tokenization.py": "Render the BPE tokenization animation used in the workshop.",
    "manim_predictive_ml.py": "Render the predictive machine learning explainer animation with Manim.",
    "manim_quantization.py": "Render the quantization explainer animation with Manim.",
}


@dataclass(slots=True)
class CommandSpec:
    label: str
    command: str


@dataclass(slots=True)
class Resource:
    kind: str
    slug: str
    title: str
    path: str
    summary: str
    tags: tuple[str, ...] = ()
    highlights: tuple[str, ...] = ()
    commands: tuple[CommandSpec, ...] = ()
    metadata: dict[str, str] = field(default_factory=dict)

    @property
    def searchable_text(self) -> str:
        parts = [
            self.kind,
            self.slug,
            self.title,
            self.path,
            self.summary,
            " ".join(self.tags),
            " ".join(self.highlights),
        ]
        return "\n".join(parts).lower()


@dataclass(slots=True)
class Catalog:
    repo_root: Path
    sections: tuple[Resource, ...]
    notebooks: tuple[Resource, ...]
    scripts: tuple[Resource, ...]
    videos: tuple[Resource, ...]

    def all_resources(self) -> tuple[Resource, ...]:
        return self.sections + self.notebooks + self.scripts + self.videos

    def by_kinds(self, *kinds: str) -> tuple[Resource, ...]:
        kind_set = set(kinds)
        return tuple(resource for resource in self.all_resources() if resource.kind in kind_set)

    def search(self, query: str, kinds: Sequence[str] | None = None) -> tuple[Resource, ...]:
        resources = self.all_resources() if kinds is None else self.by_kinds(*kinds)
        query = query.strip().lower()
        if not query:
            return tuple(resources)
        return tuple(resource for resource in resources if query in resource.searchable_text)

    @property
    def stats(self) -> dict[str, int]:
        return {
            "sections": len(self.sections),
            "notebooks": len(self.notebooks),
            "scripts": len(self.scripts),
            "videos": len(self.videos),
            "resources": len(self.all_resources()),
        }


class CatalogError(RuntimeError):
    """Raised when the workshop content can't be located."""


def _looks_like_repo_root(path: Path) -> bool:
    return (path / "README.md").exists() and (path / "notebooks").is_dir() and (path / "scripts").is_dir()


def _bundled_repo_root() -> Path | None:
    bundled = Path(__file__).resolve().parent / "assets"
    return bundled if _looks_like_repo_root(bundled) else None


def discover_repo_root(start: Path | None = None) -> Path:
    candidates: list[Path] = []
    if start is not None:
        candidates.append(start.resolve())
    candidates.append(Path.cwd().resolve())
    candidates.append(Path(__file__).resolve().parents[2])

    checked: set[Path] = set()
    for candidate in candidates:
        for path in [candidate, *candidate.parents]:
            if path in checked:
                continue
            checked.add(path)
            if _looks_like_repo_root(path):
                return path

    bundled = _bundled_repo_root()
    if bundled is not None:
        return bundled

    raise CatalogError("Could not locate the repository root from the current working directory.")


def build_catalog(repo_root: Path | None = None) -> Catalog:
    root = discover_repo_root(repo_root)
    readme_text = (root / "README.md").read_text(encoding="utf-8")
    notebook_descriptions = _parse_notebook_descriptions(readme_text)
    return Catalog(
        repo_root=root,
        sections=_parse_readme_sections(root, readme_text),
        notebooks=_parse_notebooks(root, notebook_descriptions),
        scripts=_parse_scripts(root),
        videos=_parse_videos(root),
    )


def render_resource_markdown(resource: Resource) -> str:
    lines = [f"# {resource.title}", "", resource.summary, "", f"- **Kind:** {resource.kind.title()}", f"- **Path:** `{resource.path}`"]
    if resource.tags:
        lines.append(f"- **Tags:** {', '.join(resource.tags)}")
    if resource.highlights:
        lines.extend(["", "## Highlights", *[f"- {highlight}" for highlight in resource.highlights]])
    if resource.commands:
        lines.extend(["", "## Commands", *[f"- `{command.command}`" for command in resource.commands]])
    return "\n".join(lines)


def _parse_notebook_descriptions(readme_text: str) -> dict[str, str]:
    descriptions: dict[str, str] = {}
    for match in NOTEBOOK_TABLE_PATTERN.finditer(readme_text):
        descriptions[match.group("name")] = match.group("description").strip()
    return descriptions


def _parse_readme_sections(root: Path, readme_text: str) -> tuple[Resource, ...]:
    matches = list(TOP_LEVEL_HEADING_PATTERN.finditer(readme_text))
    resources: list[Resource] = []

    for index, match in enumerate(matches):
        title = match.group("title").strip()
        if title.lower() == "license":
            continue
        start = match.end()
        end = matches[index + 1].start() if index + 1 < len(matches) else len(readme_text)
        body = readme_text[start:end].strip()
        summary = _extract_first_paragraph(body)
        highlights = tuple(line[4:].strip() for line in body.splitlines() if line.startswith("### "))
        resources.append(
            Resource(
                kind="section",
                slug=_slugify(title),
                title=title,
                path="README.md",
                summary=summary,
                tags=("readme", "workshop", "concept"),
                highlights=highlights,
                commands=(),
            )
        )

    return tuple(resources)


def _parse_notebooks(root: Path, notebook_descriptions: dict[str, str]) -> tuple[Resource, ...]:
    resources: list[Resource] = []
    for notebook_path in sorted((root / "notebooks").glob("*.ipynb")):
        notebook = json.loads(notebook_path.read_text(encoding="utf-8"))
        headings = tuple(_extract_markdown_headings(notebook)[:8])
        title = headings[0].lstrip("# ").strip() if headings else notebook_path.stem.replace("_", " ").title()
        description = notebook_descriptions.get(notebook_path.name, "Hands-on notebook for exploring the workshop topic.")
        resources.append(
            Resource(
                kind="notebook",
                slug=_slugify(notebook_path.stem),
                title=title,
                path=str(notebook_path.relative_to(root)),
                summary=description,
                tags=("lab", "notebook", "jupyter"),
                highlights=headings[1:] if len(headings) > 1 else headings,
                commands=(CommandSpec("Launch notebook", f"jupyter notebook {notebook_path.relative_to(root)}"),),
            )
        )
    return tuple(resources)


def _parse_scripts(root: Path) -> tuple[Resource, ...]:
    resources: list[Resource] = []
    for script_path in sorted((root / "scripts").glob("*.py")):
        if script_path.name == "names.py":
            continue
        summary = SCRIPT_SUMMARIES.get(script_path.name) or _humanize_script_name(script_path.stem)
        tags = ["script"]
        if script_path.name.startswith("manim_"):
            tags.append("animation")
        if script_path.name.startswith("check_"):
            tags.append("inspection")
        resources.append(
            Resource(
                kind="script",
                slug=_slugify(script_path.stem),
                title=script_path.stem.replace("_", " ").title(),
                path=str(script_path.relative_to(root)),
                summary=summary,
                tags=tuple(tags),
                commands=(CommandSpec("Run script", f"python {script_path.relative_to(root)}"),),
            )
        )
    return tuple(resources)


def _parse_videos(root: Path) -> tuple[Resource, ...]:
    resources: list[Resource] = []
    for video_path in sorted((root / "video").glob("*.mp4")):
        resources.append(
            Resource(
                kind="video",
                slug=_slugify(video_path.stem),
                title=video_path.stem.replace("_", " ").title(),
                path=str(video_path.relative_to(root)),
                summary="Rendered workshop animation ready to watch locally or embed in talks.",
                tags=("video", "animation"),
            )
        )
    return tuple(resources)


def _extract_markdown_headings(notebook: dict) -> list[str]:
    headings: list[str] = []
    for cell in notebook.get("cells", []):
        if cell.get("cell_type") != "markdown":
            continue
        text = "".join(cell.get("source", []))
        for line in text.splitlines():
            stripped = line.strip()
            if stripped.startswith("#"):
                headings.append(stripped)
                break
    return headings


def _extract_first_paragraph(text: str) -> str:
    lines: list[str] = []
    in_code_block = False
    for raw_line in text.splitlines():
        line = raw_line.strip()
        if line.startswith("```"):
            in_code_block = not in_code_block
            continue
        if in_code_block:
            continue
        if not line:
            if lines:
                break
            continue
        if line.startswith(("![", "[![", "|", ">", "### ")):
            if lines:
                break
            continue
        lines.append(line)
    if not lines:
        return "Explore this part of the workshop from the terminal app."
    return " ".join(lines)


def _humanize_script_name(script_name: str) -> str:
    words = script_name.replace("_", " ")
    return f"Run the {words} helper from the terminal lab."


def _slugify(value: str) -> str:
    value = value.lower().strip()
    value = re.sub(r"[^a-z0-9]+", "-", value)
    return value.strip("-")
